### PR TITLE
Votability and views for posts#index and posts#show

### PR DESCRIPTION
**Pull request by gingin77**
**What:** A logged in user can up vote or down vote a post on either the Posts#index or #show view. Once they have cast a vote, a new vote instance is created and the button for the vote cast is no longer clickable. Instead, the user will see the 'x' button to cancel their vote is now orange (rather than grey) and is clickable. I tried to capture all of this in the feature test - `spec/features/vote_on_posts_spec.rb`

This PR also includes a bug fix related to assigning dependent-destroy to the wrong side of the associations.

The view partials (in the shared dir because I'll eventually be using the same views to allow votes on comments) may seem confusing. I tried to name the partials based on the scenario when they are used. For example, `_down_to_up.html.erb` is used when a user has already cast a down vote, so the view hosts a clickable up vote button, but just an icon for a down vote.
votes_on_posts

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -1,8 +1,10 @@
 class PostsController < ApplicationController
+  include HackernewsClone::VoteHelper
```
Comment (zporter) [inline app/controllers/posts_controller.rb]: this appears to only be used in this controller. plans to expand in the future?

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -1,8 +1,10 @@
...
+
   before_action :authenticate_user!, only: [:create]
   before_action :authenticate_to_submit, only: :new
 
-  helper_method :post, :posts, :comment
+  helper_method :post, :posts, :comment, :new_vote
```
Comment (zporter) [inline app/controllers/posts_controller.rb]: Could move this to the module like so:

```ruby
module VoteHelper
  def self.included(base)
    base.helper_method :new_vote
  end
end
```

File: app/controllers/votes_controller.rb [hunk 3]
```diff
@@ -1,2 +1,32 @@
 class VotesController < ApplicationController
+  before_action :authenticate_user!
+  
+  helper_method :vote
+
+  def create
+    vote = current_user.votes.build(vote_params)
+    vote.save
```
Comment (zporter) [inline app/controllers/votes_controller.rb]: I would skip creating the local var here.

File: app/controllers/votes_controller.rb [hunk 4]
```diff
@@ -1,2 +1,32 @@
...
+    redirect_back(fallback_location: root_path)
+  end
+
+  def update
+    vote.value == 1 ? vote.value = -1 : vote.value = 1
```
Comment (zporter) [inline app/controllers/votes_controller.rb]: I'm curious why the toggle approach here over passing a param to set the `value`?

File: app/models/post.rb [hunk 5]
```diff
@@ -29,6 +30,10 @@ def text_post?
     text.present?
   end
 
+  def users_vote(user)
+    self.votes.find_by(user_id: user.id)
+  end
```
Comment (zporter) [inline app/models/post.rb]: could this be a `has_many` on `User`?

File: app/views/posts/_post_presenter_index.html.erb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<div>
+  <% if post.url_post? %>
+    <span><%= link_to(post.title, post.url) %></span>
+  <% else %>
+    <span><%= post.text.truncate(80, {separator: " "}) %></span>
+  <% end %>
```
Comment (zporter) [inline app/views/posts/_post_presenter_index.html.erb]: what if instead of a conditional, `post` had a `type` method that returned either `url` or `text` and that type was used to render a specific partial?  something like:

```erb
# _post_presenter_index.html.erb
<%= render partial: "#{post.type}_post_content", locals: { post: post } %>

# _url_post_content.html.erb
<span><%= link_to(post.title, post.url) %></span>

# _text_post_content.html.erb
<span><%= post.text.truncate(80, {separator: " "}) %></span>
```

A bit more flexible if more types were added in the future.

File: app/views/posts/_post_presenter_show.html.erb [hunk 7]
```diff
@@ -30,6 +30,17 @@
         <td scope="row">submitted at:</td>
         <td class="col2"><%= time_ago_in_words(post.created_at) %> ago</td>
       </tr>
+      <tr>
+        <td scope="row">vote count:</td>
```
Comment (zporter) [inline app/views/posts/_post_presenter_show.html.erb]: `vote count` is a little misleading here. the value being printed out is more of a `score`.

File: app/views/shared/vote/_active_button_options.html.erb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<% if !user_signed_in? %>
+  <%= render partial: "shared/vote/static_icons", locals: { parent: parent } %>
+<% elsif !parent.users_vote(current_user).present? %>
+  <%= render partial: "shared/vote/up_or_down", locals: { parent: parent } %>
+<% elsif parent.users_vote(current_user).value == 1 %>
+  <%= render partial: "shared/vote/up_to_down", locals: { parent: parent } %>
+<% else %>
+  <%= render partial: "shared/vote/down_to_up", locals: { parent: parent } %>
+<% end %>
```
Comment (zporter) [inline app/views/shared/vote/_active_button_options.html.erb]: these nested partials and conditionals are a little hard to follow. could the icon color be driven by a `disabled` state?  could `link_to_if` be used based on 1) whether the user is authenticated and 2) what their current vote value is for a given thing?

File: app/views/shared/vote/_edit.html.erb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<%= form_for(vote, url: vote_path(vote), action: "update") do |f| %>
+  <%= f.button type: 'submit', class: "my-button clickable", id: vote_type do %>
+    <%= tag.i class: "fa fa-chevron-circle-#{vote_type}" %>
```
Comment (zporter) [inline app/views/shared/vote/_edit.html.erb]: what is `tag` here? I've never seen this before.

File: app/views/shared/vote/_new.html.erb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<%= form_for(new_vote, url: votes_path, action: "create") do |f| %>
```
Comment (zporter) [inline app/views/shared/vote/_new.html.erb]: do `url` and `action` need to be set here?

Comment (zporter) [reply app/views/shared/vote/_new.html.erb]: replace `new_vote` with `parent.votes.build(value: one)` and you can avoid setting the `value`s in the hidden inputs below.

File: spec/features/vote_on_posts_spec.rb [hunk 11]
```diff
@@ -0,0 +1,69 @@
+require "rails_helper"
+
+feature "Vote on a post" do
+  let(:voter) { create(:oliver) }
+  let!(:url_post) { create(:url_post) }
+
+  before :each do
```
Comment (zporter) [inline spec/features/vote_on_posts_spec.rb]: `:each` doesn't need to be specified here

File: spec/features/vote_on_posts_spec.rb [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+    cast_down_vote
+    expect(posts_vote_count).to eq(-1)
+
+    expect(page).to have_css "i.not-clickable, i.fa-chevron-circle-down"
+    expect(page).to have_button("up", class: "clickable")
+    expect(page).to have_css "i.clickable, i.fa-times-circle"
+    expect(page).to have_link("delete", href: "/votes/#{Vote.last.id}")
+  end
+end
```
Comment (zporter) [inline spec/features/vote_on_posts_spec.rb:69]: worth testing an unauthenticated user?

File: spec/models/post_spec.rb [hunk 13]
```diff
@@ -1,7 +1,8 @@
 require "rails_helper"
 
 RSpec.describe Post, type: :model do
-  it { should belong_to(:submitter).dependent(:destroy) }
+  let(:post) { build(:text_post) }
+  it { expect(post).to have_many(:comments).dependent(:destroy) }
```
Comment (zporter) [inline spec/models/post_spec.rb]: nitpick: space between these 2 lines please

Comment (zporter) [reply spec/models/post_spec.rb]: line 5 could be `it { is_expected.to have_many(:comments).dependent(:destroy) }`

File: spec/models/voter_spec.rb [hunk 14]
```diff
@@ -2,9 +2,11 @@
 
 RSpec.describe Vote, type: :model do
   subject { create(:vote) }
+  let(:submitter) { build(:alice) }
+  let(:votable) { build(:text_post) }
 
-  it { should belong_to(:voter).dependent(:destroy) }
-  it { should belong_to(:voteable).dependent(:destroy) }
+  it { expect(submitter).to have_many(:votes).dependent(:destroy) }
+  it { expect(votable).to have_many(:votes).dependent(:destroy) }
```
Comment (zporter) [inline spec/models/voter_spec.rb]: I would expect these expectations to be with their respective models.

**Review by zporter — COMMENTED** [review]
left some comments. Slack me when you've responded to them.

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (gingin77) [reply app/controllers/posts_controller.rb]: yes - it'll be used by Comments

File: app/controllers/votes_controller.rb (hunk 3 above)
Comment (gingin77) [reply app/controllers/votes_controller.rb]: Thanks. I had an earlier version where I wanted some attribute from `vote` to inform the redirect path, but its not needed now.

File: app/controllers/votes_controller.rb (hunk 4 above)
Comment (gingin77) [reply app/controllers/votes_controller.rb]: I guess its one less thing to pass in locals to the `shared/vote/edit` partial, but then I have a conditional here, so hmm... 🤔

File: app/views/posts/_post_presenter_index.html.erb (hunk 6 above)
Comment (gingin77) [reply app/views/posts/_post_presenter_index.html.erb]: I like that idea

File: app/views/posts/_post_presenter_show.html.erb (hunk 7 above)
Comment (gingin77) [reply app/views/posts/_post_presenter_show.html.erb]: true

File: app/views/shared/vote/_edit.html.erb (hunk 9 above)
Comment (gingin77) [reply app/views/shared/vote/_edit.html.erb]: tag is a rails helper for HTML tags - http://api.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-tag

I needed something to support the `<i>` that font awesome uses

File: app/models/post.rb (hunk 5 above)
Comment (gingin77) [reply app/models/post.rb]: I'm not sure if this is meant to be taken at face value or is a leading question. I have now set up a `has_many(votes)` association on User. Were you suggesting I could rewrite this method as a feature of the User class? I use this method to evaluate how the vote buttons should be displayed depending on whether the current user has cast a vote for the particular post that's being evaluated. Since I'm going to see how I can simplify the conditionals I have in the views, I may not need this method, but wanted to ask about it in case there might be a better way to get the info I needed.

File: spec/features/vote_on_posts_spec.rb (hunk 12 above)
Comment (gingin77) [reply spec/features/vote_on_posts_spec.rb:69]: yeah.... see https://github.com/gingin77/hackernews_clone/issues/29 "Sort out 2 issues related to voting feature test for non-auth user"

File: app/controllers/votes_controller.rb [hunk 15]
```diff
@@ -1,2 +1,31 @@
...
+  helper_method :vote
+
+  def create
+    current_user.votes.create(vote_params)
+    redirect_back(fallback_location: root_path)
+  end
+
+  def update
+    vote.value = vote_params[:value]
+    vote.save
```
Comment (zporter) [inline app/controllers/votes_controller.rb]: could combine these 2 lines into `vote.update(value: vote_params[:value])`

File: app/models/post.rb (hunk 5 above)
Comment (zporter) [reply app/models/post.rb]: I agree, I think we can re-evaluate this comment when you've revisited the views. I have a hunch that this can actually live on the `Vote` model, since this is trying to look up a specific Vote, but won't know for sure until the dust settles on the views. Let me know if you want to pair on those at all.

File: app/controllers/comments_controller.rb [hunk 16]
```diff
@@ -1,25 +1,24 @@
...
   def new
   end
 
   def create
-    @comment = current_user.comments.build(comment_params)
-    parent = @comment.commentable
-    if @comment.save
-      redirect_to parent
+    @new_comment = current_user.comments.build(comment_params)
+    commentable = @new_comment.commentable
```
Comment (zporter) [inline app/controllers/comments_controller.rb]: I'd use the `new_comment` reader method here and below since the ivar was set above.

File: app/views/posts/show.html.erb [hunk 17]
```diff
@@ -1,5 +1,9 @@
 <%= render partial: "post_presenter_show" %>
-<%= render partial: "shared/comment/form_new", locals: { comments_parent: post, url: (post_comments_path(post)), submit_text: "Comment On Post" }  %>
+<%= render partial: "shared/comment/form_new",
```
Comment (zporter) [inline app/views/posts/show.html.erb]: can this template live under `app/views/comments`?

Comment (gingin77) [reply app/views/posts/show.html.erb]: It could. I like having it under shared because both the Posts and Comments controllers are using it, but if having a shared folder is seen as an excess, then I'd want to move both comment partials to `views/comments`, and then move everything from `shared/votes` to a `votes` dir. My thinking was that any partials that are being used by another controller would go in the `shared` dir.

Comment (zporter) [reply app/views/posts/show.html.erb]: I can dig that line of thinking 👍. I just figured the form is related to comments, so it should live under the comments dir.

File: app/controllers/comments_controller.rb [hunk 18]
```diff
@@ -1,21 +1,25 @@
 class CommentsController < ApplicationController
-  before_action :authenticate_to_submit, only: :new
+  include HackernewsClone::VoteHelper
+  include HackernewsClone::CommentHelper
```
Comment (zporter) [inline app/controllers/comments_controller.rb]: I think this is where `app/controllers/concerns` can step in to help out. something like `app/controllers/concerns/voteable_controller.rb` and `app/controllers/concerns/commentable_controller.rb` defining `VoteableController` and `CommentableController` respectively would be more inline with the Rails way™ and move away from the `Helper` suffix which might be confused with files in the `app/helpers` dir.

Comment (gingin77) [reply app/controllers/comments_controller.rb]: I made this change, then I thought about things more and I don't think the VoteHelper method belongs in a concerns dir under controllers. The method is really a query method that looks up whether a vote exists which belongs to the current user and the parent instance that a given vote button is attached to. I'm working on how to write the code so that the parent and current user get passed from the view to the vote_presenter.

File: app/controllers/concerns/voteable_controller.rb [hunk 19]
```diff
@@ -0,0 +1,11 @@
+module VoteableController
+  extend ActiveSupport::Concern
+
+  def self.included(base)
+    base.helper_method :parents_vote
+  end
+
+  def parents_vote(parent)
+    @vote = Vote.find_by(user_id: current_user, voteable_id: parent)
```
Comment (zporter) [inline app/controllers/concerns/voteable_controller.rb]: unnecessarily setting an ivar here. I'd also probably find some way to put a method on `Vote` to having something like this in your views `current_user.votes.find_by(voteable: parent)`. there might be more performant ways to go about this, but this should get the job done without requiring a method on the controllers.

Comment (gingin77) [reply app/controllers/concerns/voteable_controller.rb]: cool - thanks - closing this PR - have copied this comment to Issue https://github.com/gingin77/hackernews_clone/issues/30
